src/compute-client/src/plan/mod.rs

```diff
@@ -1674,10 +1689,53 @@ This is not expected to cause incorrect results, but could indicate a performanc
                 source.arguments.operators = Some(mfp);
             }
         }
+    }
 
-        mz_repr::explain::trace_plan(&dataflow);
+    /// Refines the plans of objects to be built as part of `dataflow` to take advantage
+    /// of monotonic operators if the dataflow refers to a single-time, i.e., is for a
+    /// one-shot SELECT query.
+    fn refine_single_time_dataflow(dataflow: &mut DataflowDescription<Self>) {
+        // Check if we have a one-shot SELECT query, i.e., a single-time dataflow.
+        if !dataflow.is_single_time() {
+            return;
+        }
 
-        Ok(dataflow)
+        // Upgrade single-time plans to monotonic.
+        for build_desc in dataflow.objects_to_build.iter_mut() {
+            let mut todo = vec![&mut build_desc.plan];
+            while let Some(expression) = todo.pop() {
+                match expression {
+                    Plan::Reduce { plan, .. } => {
+                        use crate::plan::reduce::CollationPlan;
+                        // Upgrade non-monotonic hierarchical plans to monotonic with mandatory consolidation.
+                        match plan {
+                            ReducePlan::Collation(CollationPlan { hierarchical, .. }) => {
```

Comment:
nit: I think you can define a `flag_snapshot` (name t.b.d.) for `CollationPlan` similar to `Hierarchical` where you put this logic.

Reply:
Agreed, done!